include/bluetooth/services/ddfs.h

```diff
@@ -0,0 +1,261 @@
+	DDFS_EVT_AZIMUTH_MEAS_NOTIFY_DISABLE,
+
+	/** Elevation Measurement notification enable. */
+	DDFS_EVT_ELEVATION_MEAS_NOTIFY_ENABLE,
+
+	/** Elevation Measurement notification disable. */
+	DDFS_EVT_ELEVATION_MEAS_NOTIFY_DISABLE,
+};
+
+/**@brief Direction and Distance Service event handler type. */
+typedef void (*bt_ddfs_evt_handler) (enum bt_ddfs_evt evt);
+
+/** @brief Bitmask set of supported features. */
+struct bt_ddfs_features {
+	/** Distance Measurement RTT ranging mode supported bit. */
+	uint8_t ranging_mode_rtt : 1;
+
+	/** Distance Measurement MCPD ranging mode supported bit. */
+	uint8_t ranging_mode_mcpd : 1;
+};
+
+/** @brief Distance Measurement configuration parameters. */
+struct ddfs_dm_config {
+	/** Ranging mode */
+	enum ddfs_dm_ranging_mode mode;
+};
+
+/** @brief Control Point callback structure.*/
+struct bt_ddfs_cp_cb {
+	/** @brief Set Ranging mode for Distance Measurement.
+	 *
+	 * @param mode New ranging mode for distanance measurement.
+	 * @retval 0 If the operation was successful.
+	 *         Otherwise, a negative error code is returned.
+	 */
+	int (*set_dm_ranging_mode)(uint8_t mode);
+
+	/** @brief Read Distance Measurement configuration.
+	 *
+	 * @param config Pointer to configuration object.
+	 * @retval 0 If the operation was successful.
+	 *         Otherwise, a negative error code is returned.
+	 */
+	int (*read_dm_config)(struct ddfs_dm_config *config);
+};
+
+struct bt_ddfs_distance_measurement {
+	/* Quality indicator */
+	enum ddfs_quality quality;
+
+	/* Bluetooth LE Device Address */
+	bt_addr_le_t bt_addr;
+
+	/* Mode used for ranging */
+	enum ddfs_dm_ranging_mode ranging_mode;
+	union {
+		struct {
+			/* MCPD: Distance estimate based on IFFT of spectrum */
+			uint16_t ifft;
+
+			/* MCPD: Distance estimate based on average phase slope estimation */
+			uint16_t phase_slope;
+
+			/* RSSI: Distance estimate based on Friis path loss formula */
+			uint16_t rssi_openspace;
+
+			/* Best effort distance estimate */
+			uint16_t best;
+		} mcpd;
+		struct {
+			/* RTT: Distance estimate based on RTT measurement */
+			uint16_t rtt;
+		} rtt;
+	} dist_estimates;
+};
+
+struct bt_ddfs_azimuth_measurement {
+	/* Quality indicator */
+	enum ddfs_quality quality;
+
+	/* Bluetooth LE Device Address */
+	bt_addr_le_t bt_addr;
+
+	/* Azimuth estimate */
+	uint16_t value;
+};
+
+struct bt_ddfs_elevation_measurement {
+	/* Quality indicator */
+	enum ddfs_quality quality;
+
+	/* Bluetooth LE Device Address */
+	bt_addr_le_t bt_addr;
+
+	/* Elevation estimate */
+	int8_t value;
+};
+
+/** @brief Direction and Distance Finding Service initialization parameters. */
+struct bt_ddfs_init_params {
+	/** Initial value for features. */
+	struct bt_ddfs_features dm_features;
+
+	/** Event handler to be called for handling events in the DDF Service. */
+	bt_ddfs_evt_handler evt_handler;
+
+	/** Control Point callback structure.*/
+	const struct bt_ddfs_cp_cb *cp_cb;
```

Comment:
Does it make sense to have two different callback structures?

You could merge them into one and use a proper prefix to indicate when the callback corresponds to the CP action or notification action. Unnecessary callbacks can be initialized to null